flybywiresim/a32nx
# refactor: introduce new SimVar hooks for React

Fixes #3374 Fixes #3408 

## Summary of Changes


This pull request introduces the new SimVar utilities for React co-authored by @theomessin and @beheh.

It introduces the following 3 new hooks:
- **`useSimVar`**
  The basic version. You'll want this 90% of the time.
- **`useGlobalVar`**
  The same, but for global variables (like `ZULU TIME`).
- **`useInteractionSimVar`**  
  The optimized version. If you emit an H event anyway whenever it changes (e.g. a knob), this is for you.
- **`useSplitSimvar`**
  Use this if you're accessing a SimVar that is set using K events (mostly simulator SimVars). Not used often.

Here's a few examples:
```typescript
// the return value is the value itself and a setter, similar to useState
const [v1, setV1] = useSimVar('L:AIRLINER_V1_SPEED', 'Knots');

// only refresh the SimVar every 500ms (unless this SimVar is lower elsewhere)
const [lightsTest] = useSimVar('L:XMLVAR_LTS_Test', 'Bool', 500);

// access a GlobalVar which is refreshed at least every 100ms
const time = useGlobalVar('ZULU TIME', 'seconds', 100);

// the XML updates the SimVar and emits an H event, so we can use the optimized version
const [toggleSwitch, setToggleSwitch] = useInteractionSimVar(
  'L:A32NX_RMP_L_TOGGLE_SWITCH', 'bool',
  'A32NX_RMP_L_TOGGLE_SWITCH'
);

// read the SimVar 'COM STANDBY FREQUENCY:2', and set it through 'K:COM_2_RADIO_SET_HZ'
const [frequencyTwo, setFrequencyTwo] = useSplitSimVar(
  'COM STANDBY FREQUENCY:2', 'Hz',
  'K:COM_2_RADIO_SET_HZ', 'Hz'
);
```

We've also decided to write in Typescript, so we've moved the `tsconfig.json` and had to make a tiny change to `rollup.config.js`.

Other than that, we've refactored the RMP to use the new hooks as a live example of how these new SimVars work and to ensure their API is useable.

## Additional context

Future work outside of the scope of this PR:
- Add a `<RefreshRate>` component to set a default refresh rate in parts of the tree (rather than defaulting to 0)
- Add a way to guarantee SimVar updates, like a number on a screen only changing every 500ms (note that maxInterval is only best-effort for now)

## Testing instructions

- Test that the Radio Management Panels work as they should.
- VHF1, VHF2, VHF3 are split and should fully work.
- All states should load with correct frequencies.
- Ensure that any previously reported RMP bugs have been resolve (#3302, #3362, #3374) , 
- Test whether there are any performance issues that need to be addressed (reactivity, fps).
- Make sure to perform long flights.
- [!] The RMP Nav modes do not work, this is expected.
- [!] The RMP LED lights do not light up when in lights test mode, this will be a separate fix PR.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link
